### PR TITLE
Add multi-grad hooks

### PR DESCRIPTION
**Pull request by soulitzer**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #85849
* #79056
* __->__ #86260
* #85711

### CODE REVIEW

File: torch/csrc/Module.cpp [hunk 1]
```diff
@@ -731,30 +731,25 @@ PyObject* THPModule_willEngineExecuteNode(PyObject* _unused, PyObject* arg) {
...
           "currently running autograd.grad(). This is currently not supported.");
-      Py_RETURN_TRUE;
     }
   }
-
+  if (ret) {
+    Py_RETURN_TRUE;
+  } else {
+    Py_RETURN_FALSE;
+  }
```
Comment (soulitzer) [inline torch/csrc/Module.cpp:752]: If we pass the node not part of the graph as input=, exec_info will still think we should execute. So now we update to always check whether the node is in the graph.

File: torch/csrc/Module.cpp [hunk 2]
```diff
@@ -731,30 +731,25 @@ PyObject* THPModule_willEngineExecuteNode(PyObject* _unused, PyObject* arg) {
...
+  const auto nodes_in_graph =
+      torch::autograd::get_current_graph_task_nodes_in_graph();
+  bool ret = nodes_in_graph->find(node) != nodes_in_graph->end();
+  if (ret && !exec_info->empty()) {
     auto it = exec_info->find(node);
     if (it == exec_info->end() || !it->second.should_execute()) {
-      Py_RETURN_FALSE;
+      ret = false;
     } else {
       THPUtils_assert(
```
Comment (albanD) [inline torch/csrc/Module.cpp:742]: not necessarily for this PR but we should move this to use TORCH_CHECK

**Review by albanD — COMMENTED** [review]
base sounds good.
Only some question on the multithreading side

File: test/test_autograd.py [hunk 3]
```diff
@@ -9019,19 +9052,38 @@ def train_fn_grad(x):
...
+                self.assertEqual(res, grad_is_none)
 
-        def registers_hooks_for_each_thread():
-            with torch.autograd.graph.saved_tensors_hooks(pack, lambda x: x):
-                x = torch.ones(5, 5, requires_grad=True)
-                with warnings.catch_warnings(record=True) as w:
-                    y = x * x
-                    # should raise two warnings from x being saved twice
-                    self.assertEqual(len(w), 2)
-            y.sum().backward()
```
Comment (soulitzer) [inline test/test_autograd.py:9034]: I removed this because it didn't actually run any code previously, and I tried modifying it to try to get it work, but the number of warnings produced seems to be non-deterministic.

```python
def test_multithread_saved_tensors_hooks(self):
      def pack(x):
          warnings.warn("pack")
          return x

      num_warnings = [0]
      def registers_hooks_for_each_thread():
          warnings.simplefilter("always")
          with torch.autograd.graph.saved_tensors_hooks(pack, lambda x: x):
              x = torch.ones(5, 5, requires_grad=True)
              with warnings.catch_warnings(record=True) as w:
                  y = x * x
              num_warnings[0] += len(w)

      with warnings.catch_warnings(record=True) as w:
          self._run_py_multithread_fn(registers_hooks_for_each_thread, num_threads=10)

      self.assertEqual(num_warnings[0] + len(w), 20)
```

File: test/test_autograd.py [hunk 4]
```diff
@@ -9019,19 +9070,38 @@ def train_fn_grad(x):
         # be accumulate to the same place and should be the same
         self._run_py_multithread_fn(train_fn_grad, (x,))
 
-    def test_multithread_saved_tensors_hooks(self):
-        def pack(x):
-            warnings.warn("pack")
-            return x
+    def test_multi_grad_hooks(self):
+        # Multihooks should behave independently per execution of backward
```
Comment (albanD) [inline test/test_autograd.py:9074]: I think a good way to test this is to use a custom Function that raise an error.
That way you will leave a multi-hook partially applied.

File: torch/autograd/graph.py [hunk 5]
```diff
@@ -166,3 +174,98 @@ def disable_saved_tensors_hooks(error_message):
...
+        else:
+            return t.grad_fn
+
+    grad_fns = list(map(get_grad_fn, tensors))
+
+    def get_inner_hook(idx):
+        def inner_hook(grad: torch.Tensor):
+            nonlocal count, nb_calls, buffer
+            id = torch._C._current_graph_task_id()
+            assert id != -1
```
Comment (albanD) [inline torch/autograd/graph.py:234]: avoid plain assert please

**Review by albanD — APPROVED** [review]
SGTM
Wondering if we could make sure we don't leak the state when an error occurs in backward :thinking: 
But can be done in a follow up

**Comment by soulitzer** [comment]
> Wondering if we could make sure we don't leak the state when an error occurs in backward 🤔

I'm not sure how much we should be worried about the buffer of gradients living longer than than necessary because the lifetime of the buffer is tied to that of the graph. Is there a use case where an error occurs in backward, but the user chooses to run backward on the same graph again without trying to rerun forward?

Ideal solution, but requires a new API:
- (1) To be able to reset the state of the hook as soon as backward errors out, we'd need the engine to catch the errors on backward and allow us to register a hook to interpose at that time. Maybe this API could be generally useful for other clean up?

Other solution:
- (2) have the user manually remove and reregister the hook to clear the buffer (maybe this could be OK if we document it)
- (3) catch this invalid state the next time someone runs backward on the same graph (need to check reentrancy + graph_task id + thread id)

Solution (2) seems like the simplest, so I'd probably prefer that unless there are other use cases for the API required for (1).
